s2/gen.go

```diff
@@ -0,0 +1,1590 @@
+	if retval != nil {
+		ADDQ(U8(2), retval)
+	}
+	ADDQ(U8(2), dstBase)
+	JMP(LabelRef("memmove_" + name))
+
+	Label("one_byte_" + name)
+	SHLB(U8(2), n.As8())
+	MOVB(n.As8(), Mem{Base: dstBase})
+	if retval != nil {
+		ADDQ(U8(1), retval)
+	}
+	ADDQ(U8(1), dstBase)
+	// Fallthrough
+
+	Label("memmove_" + name)
+
+	// copy(dst[i:], lit)
+	if true {
+		dstEnd := GP64()
+		if updateDst {
+			LEAQ(Mem{Base: dstBase, Index: litLen, Scale: 1}, dstEnd)
+		}
+		length := GP64()
+		MOVL(litLen.As32(), length.As32())
+		genMemMove2("emit_lit_memmove_"+name, dstBase, litBase, length, end, avx)
+		if updateDst {
+			MOVQ(dstEnd, dstBase)
+		}
+	} else {
+		genMemMove("emit_lit_memmove_"+name, dstBase, litBase, litLen, end)
+	}
+	return
+}
+
+// genEmitRepeat generates a standlone emitRepeat.
+func genEmitRepeat() {
+	TEXT("emitRepeat", NOSPLIT, "func(dst []byte, offset, length int) int")
+	Doc("emitRepeat writes a repeat chunk and returns the number of bytes written.",
+		"Length must be at least 4 and < 1<<32", "")
+	Pragma("noescape")
+
+	dstBase, offset, length, retval := GP64(), GP64(), GP64(), GP64()
+
+	// retval = 0
+	XORQ(retval, retval)
+
+	Load(Param("dst").Base(), dstBase)
+	Load(Param("offset"), offset)
+	Load(Param("length"), length)
+	emitRepeat("standalone", length, offset, retval, dstBase, LabelRef("gen_emit_repeat_end"))
+	Label("gen_emit_repeat_end")
+	Store(retval, ReturnIndex(0))
+	RET()
+}
+
+// emitRepeat can be used for inlining an emitRepeat call.
+// length >= 4 and < 1<<32
+// length is modified. dstBase is updated. retval is added to input.
+// retval can be nil.
+// Will jump to end label when finished.
+// Uses 1 GP register.
+func emitRepeat(name string, length, offset, retval, dstBase reg.GPVirtual, end LabelRef) {
+	Label("emit_repeat_again_" + name)
+	tmp := GP32()
+	MOVL(length.As32(), tmp) // Copy length
+	// length -= 4
+	LEAL(Mem{Base: length, Disp: -4}, length.As32())
+
+	// if length <= 4 (use copied value)
+	CMPL(tmp.As32(), U8(8))
+	JLE(LabelRef("repeat_two_" + name))
+
+	// length < 8 && offset < 2048
+	CMPL(tmp.As32(), U8(12))
+	JGE(LabelRef("cant_repeat_two_offset_" + name))
+	CMPL(offset.As32(), U32(2048))
+	JLT(LabelRef("repeat_two_offset_" + name))
+
+	const maxRepeat = ((1 << 24) - 1) + 65536
+	Label("cant_repeat_two_offset_" + name)
+	CMPL(length.As32(), U32((1<<8)+4))
+	JLT(LabelRef("repeat_three_" + name)) // if length < (1<<8)+4
+	CMPL(length.As32(), U32((1<<16)+(1<<8)))
+	JLT(LabelRef("repeat_four_" + name)) // if length < (1 << 16) + (1 << 8)
+	CMPL(length.As32(), U32(maxRepeat))
+	JLT(LabelRef("repeat_five_" + name)) // If less than 24 bits to represent.
+
+	// We have have more than 24 bits
+	// Emit so we have at least 4 bytes left.
+	LEAL(Mem{Base: length, Disp: -(maxRepeat - 4)}, length.As32()) // length -= (maxRepeat - 4)
+	MOVW(U16(7<<2|tagCopy1), Mem{Base: dstBase})                   // dst[0] = 7<<2 | tagCopy1, dst[1] = 0
+	MOVW(U16(65531), Mem{Base: dstBase, Disp: 2})                  // 0xfffb
```

Comment:
I personally think this kind of thing might read nicer if you have a line earlier setting up:

```go
dst := Mem{Base: dstBase}
```

then instead of `Mem{Base: dstBase, Disp: 2}` you can do `dst.Offset(2)`.

Reply:
It becomes very verbose when you separate everything with very little functionality on screen at one time.